trieste/acquisition/function.py

```diff
@@ -852,3 +853,202 @@ def batch_ei(at: TensorType) -> TensorType:
+    by successively maximizing function penalized around alread selected points.
+    Talk about penalization being multiplicative (do log space)
+
+    For our penalization function, we implement the soft penalization strategy of :cite:`Gonzalez:2016`, 
+    where an estimate of the objective function's Lipschitz constant is used to control the size 
+    of penalization. The Lipschitz constant and additionl penalization parameters are estimated once 
+    when preparing the :const:`AcquisitionFunctionMaker` and reused for each 
+    individual :const:`AcquisitionFunction`) it makes.
+    
+    """
+
+
+    def __init__(self, search_space: SearchSpace, grid_size: int = 500):
+        """
+        :param search_space: The global search space over which the optimisation is defined.
+        :param grid_size: Size of the grid over which the Lipschitz constant is estimated. We recommend
+            scaling this with search space dimension.
+        """
+        self._search_space = search_space
+        if grid_size <= 0:
+            raise ValueError(f"grid_size must be positive, got {grid_size}")
+        self._grid_size = grid_size
+
+        self._lipschitz_constant = None
+        self._eta = None
+        
+
+    def prepare_acquisition_function_maker(
+            self, dataset: Dataset, model: ProbabilisticModel, pending_points: TensorType,
+    ) -> AcquisitionFunction:
+        """
+        :param dataset: The data from the observer.
+        :param model: The model over the specified ``dataset``.
+        :param pending_points: The points we penalize with respect to.	
+        :return: The (log) expected improvement acqusiiton function penalized with respect to the pending points.
+        """
+        if len(dataset.query_points) == 0:
+            raise ValueError("Dataset must be populated.")
+
+        if not pending_points: # only compute penalization parameters once per optimization step
+
+	        samples = self._search_space.sample(num_samples=self._grid_size)
+	        samples  = tf.concat([dataset.query_points, samples ], 0)
+
+	        def get_lipschitz_estimate(sampled_points)-> tf.Tensor: # use max norm of posterior mean gradients
+	            with tf.GradientTape() as g: 
+	                g.watch(sampled_points)
+	                mean, _ = model.predict(sampled_points)
+	            grads = g.gradient(mean,sampled_points)
+	            grads_norm =  tf.norm(grads, axis=1)
+	            max_grads_norm = tf.reduce_max(grads_norm)
+	            return max_grads_norm
+
+	        lipschitz_constant = get_lipschitz_estimate(samples)
+
+	        if lipschitz_constant < 1e-5: # threshold to improve numerical stability for 'flat' models
+	                lipschitz_constant = 10
+
+	        eta = tf.reduce_min(mean, axis=0)
+
+	        self._lipschitz_constant = lipschitz_constant
+	        self._eta = eta
+
+
+        log_base_acquisition = tf.math.log(expected_improvement(model, eta))
```

Comment:
```suggestion
        log_base_acquisition = tf.math.log(expected_improvement(model, self._eta))
```

and maybe move below the not pending_points check